### PR TITLE
[WIP] Continuous integration using Github Actions

### PR DESCRIPTION
**Pull request by katyo**
Implemented tasks:
- [x] Check code formatting rules
- [x] Check code quality using `clippy`
- [x] Generate bindings for supported platforms (stable Rust)
- [x] Check building on all supported platforms (stable Rust)
- [x] Check building on different Rust channels (stable/beta/nightly, Linux x86_64)
- [x] Run tests on native Linux platforms (x86_64 and i686)
- [x] Create pull-requests for updating generated bindings
- [x] Publish crates when pushing version tag

All tasks runs on ubuntu 20.04.
To get publishing working you need set `CRATES_TOKEN` secret.

Currently supported prebuilt bindings:
- Linux x86
- Linux x86_64
- Linux arm
- Linux aarch64

Blocked by [drm-rs#74](/Smithay/drm-rs/pull/74)

### CODE REVIEW

File: gbm-sys/build.rs
Comment (MarijnS95) [inline gbm-sys/build.rs:71]: I think this is the default for [`File::create`](https://doc.rust-lang.org/std/fs/struct.File.html#method.create):

```suggestion
            let mut env_file = File::create(github_env).unwrap();
```

File: src/buffer_object.rs
Comment (MarijnS95) [inline src/buffer_object.rs:468]: `f.write_str("the string")` might be nicer, though that same function should be used under the hood when calling `Display::fmt` on the string.

**Review by MarijnS95 — CHANGES_REQUESTED** [review]
Now that https://github.com/Smithay/drm-rs/pull/74 is merged this is unblocked, right? Can it be rebased and should the `drm-rs` branch be changed to `develop` or should this wait until `0.4` is out?

Should this be merged to `master` or the `develop` branch? What's the difference, given that both are working branches that have yet to be released?

@katyo and some commits based on https://github.com/Smithay/drm-rs/pull/81 if you don't mind.

**Comment by katyo** [comment]
@MarijnS95 @Drakulix I plans to rebase (and rewrite) this PR on `develop` branch after merging #80, #81 and #82 to `drm-rs`.

**Comment by MarijnS95** [comment]
@katyo Sounds good, unfortunately all those PRs are blocked by `nightly` failing to compile `clang-sys`, should be fixed by https://github.com/rust-lang/rust/pull/84130.

When you do, would you mind running the example in rustdoc through `rustfmt`, and then copying it back to the README? Thanks!

**Comment by MarijnS95** [comment]
@katyo And they're all in now since nightly has been fixed :)

**Comment by katyo** [comment]
Seems I need relaunch this pr from the scratch, because a lot of changes happened in master.
@Drakulix Should I rebase it to `develop` branch instead of `master`?

**Comment by Drakulix** [comment]
> Seems I need relaunch this pr from the scratch, because a lot of changes happened in master.
> @Drakulix Should I rebase it to `develop` branch instead of `master`?

Yeah sorry, the new release of drm-rs, usage of drm-fourcc and some general updates (including adding some new gbm-functions) produced a lot of changes.

Please target `develop`. I'll plan to merge that into master again eventually once new functionalities are all properly tested, but that may take some more time.

File: gbm-sys/build.rs
Comment (katyo) [reply gbm-sys/build.rs:71]: We do not need `truncate(true)`, which is also default for `File::create`.

**Comment by katyo** [comment]
Seems something went wrong on 32-bit targets.

**Comment by katyo** [comment]
@Drakulix All tests passed for now except minimum versions check.

**Comment by katyo** [comment]
Unfortunately I does not understand how check minimal-versions actually works.

It does not fails on my machine. As I see `wayland-sys` v0.28.5 depends from `pkg-config` ^0.3.7.

On my machine `cargo +nightly tree --all --all-features --all-targets -Z minimal-versions` outputs the following (clipped):
```
└── wayland-sys v0.28.5
        [build-dependencies]
        └── pkg-config v0.3.19
```

But on CI `pkg-config` v0.3.0 is used instead. Why is it happens?

**Comment by MarijnS95** [comment]
@katyo Remove your `Cargo.lock`, then it'll show the same :)

**Comment by katyo** [comment]
@MarijnS95 Is it means that cargo does not follows semver specs?
[The 0.7.0 does not satisfy constraint ^0.7.3](https://jubianchi.github.io/semver-check/#/^0.7.3/0.7.0).

**Comment by MarijnS95** [comment]
@katyo Not exactly sure what you mean. `wayland-sys-0.28.0` has a dependency on `pkg-config` `0.3`. Cargo uses [semver caret matching by default](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio) meaning that this allows any `pkg-config` version `>= 0.3.0`.

Where are you seeing versions of `0.7.0` and `0.7.3`?

**Comment by katyo** [comment]
Hmm, seems I forget include `wayland-server` version fix into commit.

**Comment by MarijnS95** [comment]
@katyo Neat, reminder that you don't need `^` though. That's implied by `cargo`.

Doesn't look like any GH Actions ran now? There's only the failed Travis run...

**Comment by katyo** [comment]
@MarijnS95 It because the PR doesn't merged. See the actions tab on my fork. https://github.com/katyo/gbm.rs/actions/runs/782178801

**Comment by MarijnS95** [comment]
@katyo Indeed that's how I remember it working (actions have to be merged to the default branch, or a non-fork branch first before anything runs), but somehow recall seeing a successful actions run in here. Not that it really matters though.

I guess this and #7 are now ready for merging, and perhaps a branch cleanup and crates release is nearby?

**Comment by Drakulix** [comment]
> @katyo Indeed that's how I remember it working (actions have to be merged to the default branch, or a non-fork branch first before anything runs), but somehow recall seeing a successful actions run in here. Not that it really matters though.
> 
> I guess this and #7 are now ready for merging, and perhaps a branch cleanup and crates release is nearby?

I will take the time to review (and likely merge) this in the next couple of days and yes, I plan to do a drm-rs and gbm-rs release shortly afterwards.

**Comment by MarijnS95** [comment]
@katyo apologies, #7 introduced some conflicts. Should be easy for you to solve by simply rerunning rustfmt (at least simpler than having to reapply comment changes on top of rustfmt :grimacing:). Thanks!

File: .github/workflows/ci.yml
Comment (Drakulix) [inline .github/workflows/ci.yml:71]: I think we do not need nightly here, please migrate `check` and `check-minimal` to stable.

File: gbm-sys/build.rs
Comment (Drakulix) [reply gbm-sys/build.rs:138]: In that case, please use `File::create`

File: src/lib.rs
Comment (Drakulix) [inline src/lib.rs:73]: Please use `, 32, 32` as arguments here

**Review by Drakulix — CHANGES_REQUESTED** [review]
Thanks, I found some small nitpicks, after overall, this is again really solid. :+1:

File: .github/workflows/ci.yml
Comment (katyo) [reply .github/workflows/ci.yml:71]: I used `nightly` for `check` because it supports more rich set of checks.
IIRC the `-Z minimal-versions` is supported by `nightly` only so it required for `check-minimal`.

File: src/lib.rs
Comment (katyo) [reply src/lib.rs:73]: We need just append variable assignment line at end of `$GITHUB_ENV` file.
The `File::create` truncates existing file, which is wrong usage.

File: .github/workflows/ci.yml
Comment (Drakulix) [reply .github/workflows/ci.yml:71]: True, we need it for check-minimal, well then it is fine for me.

File: src/lib.rs
Comment (Drakulix) [reply src/lib.rs:73]: Weird, me neither. In that case, the File::create discussion is also resolved for me.

Comment (MarijnS95) [reply src/lib.rs:73]: > This reply addressed to comment above, seems I can't reply to it.

That is because [that comment](https://github.com/Smithay/gbm.rs/pull/4#discussion_r624532695) was sent in reply to [my comment](https://github.com/Smithay/gbm.rs/pull/4#discussion_r612706392) in [my review](https://github.com/Smithay/gbm.rs/pull/4#pullrequestreview-634925648), as part of [Drakulix' review](https://github.com/Smithay/gbm.rs/pull/4#pullrequestreview-649800089). GitHub does not really know what to do in that case, and simply renders it twice. Once in Drakulix' review without reply field, the second time as part of the original "comment thread". You can easily jump to the original by clicking the "xxx hours ago", that gives a unique permalink to the comment-in-thread (where you can reply) for both of the rendered instances.

File: .github/workflows/ci.yml
Comment (Drakulix) [inline .github/workflows/ci.yml:7]: Please remove :)

**Review by Drakulix — APPROVED** [review]
Looks good to go!
One last cleanup step and then we hit that Merge button

File: .github/workflows/ci.yml
Comment (MarijnS95) [reply .github/workflows/ci.yml:116]: @katyo It doesn't look like this item was given a friendly name.

By the way, for all the other review comments (with "outdated" tag on them), please mark them as resolved (despite this PR already been merged), as the reviewer I cannot even close my own comments :(

**Pull request merged**
